src/draft06.rs

```diff
@@ -0,0 +1,756 @@
+#[derive(Debug, Clone)]
+pub struct ObliviousDoHConfig {
+    version: u16,
+    length: u16,
+    contents: ObliviousDoHConfigContents,
+}
+
+impl Serialize for &ObliviousDoHConfig {
+    fn serialize<B: BufMut>(self, buf: &mut B) -> Result<()> {
+        buf.put_u16(self.version);
+        buf.put_u16(self.length);
+        self.contents.serialize(buf)
+    }
+}
+
+impl Deserialize for ObliviousDoHConfig {
+    fn deserialize<B: Buf>(mut buf: &mut B) -> Result<Self> {
+        if buf.remaining() < 2 {
+            return Err(Error::ShortInput);
+        }
+        let version = buf.get_u16();
+        let mut contents = read_lengthed(&mut buf)?;
+        let length = contents.len() as u16;
+
+        Ok(Self {
+            version,
+            length,
+            contents: parse(&mut contents)?,
+        })
+    }
+}
+
+impl From<ObliviousDoHConfig> for ObliviousDoHConfigContents {
+    fn from(c: ObliviousDoHConfig) -> Self {
+        c.contents
+    }
+}
+
+impl From<ObliviousDoHConfigContents> for ObliviousDoHConfig {
+    fn from(c: ObliviousDoHConfigContents) -> Self {
+        Self {
+            version: ODOH_VERSION,
+            length: c.len() as u16,
+            contents: c,
+        }
+    }
+}
+
+/// Contains the HPKE suite parameters and the
+/// resolver (target's) public key.
+#[derive(Debug, Clone)]
+pub struct ObliviousDoHConfigContents {
+    kem_id: u16,
+    kdf_id: u16,
+    aead_id: u16,
+    // protocol: length prefix
+    public_key: Bytes,
+}
+
+impl ObliviousDoHConfigContents {
+    /// Creates a KeyID for an `ObliviousDoHConfigContents` struct
+    fn identifier(&self) -> Result<Vec<u8>> {
+        let buf = compose(self)?;
+
+        let key_id_info = LABEL_KEY_ID.to_vec();
+        let prk = Hkdf::<<Kdf as KdfTrait>::HashImpl>::new(None, &buf);
+        let mut key_id = [0; KDF_OUTPUT_SIZE];
+        prk.expand(&key_id_info, &mut key_id)
+            .map_err(|_| Error::from(HpkeError::InvalidKdfLength))?;
+        Ok(key_id.to_vec())
+    }
+
+    fn len(&self) -> usize {
+        2 + 2 + 2 + 2 + self.public_key.len()
+    }
+}
+
+impl Serialize for &ObliviousDoHConfigContents {
+    fn serialize<B: BufMut>(self, buf: &mut B) -> Result<()> {
+        buf.put_u16(self.kem_id);
+        buf.put_u16(self.kdf_id);
+        buf.put_u16(self.aead_id);
+
+        buf.put_u16(to_u16(self.public_key.len())?);
+        buf.put(self.public_key.clone());
+        Ok(())
+    }
+}
+
+impl Deserialize for ObliviousDoHConfigContents {
+    fn deserialize<B: Buf>(mut buf: &mut B) -> Result<Self> {
+        if buf.remaining() < 2 + 2 + 2 {
+            return Err(Error::ShortInput);
+        }
+
+        let kem_id = buf.get_u16();
+        let kdf_id = buf.get_u16();
+        let aead_id = buf.get_u16();
+
+        if kem_id != KEM_ID || kdf_id != KDF_ID || aead_id != AEAD_ID {
+            return Err(Error::InvalidParameter);
+        }
+
+        let public_key = read_lengthed(&mut buf)?;
```

Comment:
Should we also validate the public key length (32B)?

Reply:
@xofyarg can we please fix this in a separate followup PR?

Reply:
Sure. I'll submit another one later.